src/poly/evaluator.rs

```diff
@@ -0,0 +1,386 @@
+use std::{
+    collections::HashMap,
+    hash::Hash,
+    marker::PhantomData,
+    ops::{Add, Mul, Neg, Sub},
+};
+
+use group::ff::Field;
+use pasta_curves::arithmetic::FieldExt;
+
+use super::{
+    Basis, Coeff, EvaluationDomain, ExtendedLagrangeCoeff, LagrangeCoeff, Polynomial, Rotation,
+};
+use crate::arithmetic::parallelize;
+
+/// A reference to a polynomial registered with an [`Evaluator`].
+#[derive(Clone)]
+pub(crate) struct AstLeaf<E, Id, B: Basis> {
+    id: Id,
+    rotation: Rotation,
+    _evaluator: PhantomData<(E, B)>,
+}
+
+impl<E, Id: Copy, B: Basis> AstLeaf<E, Id, B> {
+    /// Produces a new `AstLeaf` node corresponding to the underlying polynomial at a
+    /// _new_ rotation. Existing rotations applied to this leaf node are ignored and the
+    /// returned polynomial is not rotated _relative_ to the previous structure.
```

Comment:
THis is kind-of odd, because it depends on how the original `AstLeaf` was constructed rather than what it currently represents, which is a leaky abstraction. Applying a rotation relative to the previous one would solve that issue (I think that in practice we're normally calling `with_rotation` on unrotated nodes, so it would still work).

Reply:
I originally proposed making this a relative rotation, but @ebfull disagreed. I don't recall his exact rationale.

Reply:
We could alternatively replace this API with a registration API that takes a rotation, and under the hood ensures the `AstLeaf` references the same polynomial of this API is called alongside the no-rotation API. There will definitely be contexts where we can't access both the evaluator and the original polynomial, but I don't know how often we will need to get a rotation of an `AstLeaf` there rather than at the beginning.

Reply:
I am fine with an API where you start with something that hasn't been rotated and then get to rotate it, but an API that allows you to rotate on top of a already-rotated leaf is annoying because it requires us to reduce modulo `2^k`, which requires passing the `k` constant around all over the place.

Reply:
Why would we need to reduce modulo 2<sup>k</sup>? We're representing an element of Z<sub>2<sup>k</sup></sub> as an integer, which is a valid representation even if it's not canonical. The only weirdness is that equivalent rotations will not compare equal, but if I understand correctly that's not a problem: it could only affect efficiency.

In practice I think it will also not affect efficiency, because practical circuits will only use small rotations (except possibly for [when we span columns](https://zcash.github.io/halo2/design/proving-system/permutation.html#spanning-a-large-number-of-columns) in the permutation argument, which requires a rotation of u = 2k - t - 1, although we could write that as - t - 1).